### PR TITLE
Add SharingReport tooling for thesis evaluation (Phase 1)

### PR DESCRIPTION
**Pull request by timotejAdamec**
> **Stacked on #226** — features-shared top-level restructure, which is itself stacked on #224 (sync refactor). Merge chain: `main ← #224 ← #226 ← #223`. As each upstream PR merges GitHub auto-retargets the next one's base.

## Problem Statement

The thesis *"Multiplatform snagging system with code sharing maximisation"* needs a new Evaluation chapter (Kapitola 4: Vyhodnocení) backed by **measured** data — sharing distribution across modules/layers/source sets and ripple-effect measurements from evolvability experiments. Today the thesis has no measurement tooling; every number would have to be hand-counted.

Phase 1 of the evaluation plan builds the **data pipeline** that every later phase depends on: §4.2 sharing numbers, §4.3 ripple classification (dependency closure consumes the same metadata), §4.4 Wear OS ripple measurement, and every figure in Kapitola 4. No prose is written until the CSVs exist.

## Solution

Three-stage pipeline — each stage independently runnable and testable:

1. **`./gradlew sharingReport`** — new `SharingReportTask` registered by the `snagSharingReport` convention plugin (applied to the root project). Walks `rootProject.subprojects` via `gradle.projectsEvaluated`, reuses the existing `parseModulePath()` from `build-logic/.../architecture/`, probes the nine canonical Snag convention plugins plus `com.android.application` as a marker, scans each project's `src/*/` directories, and emits one CSV row per `(module × source set)` pair to `build/reports/sharing/sharing_report.csv`. No LOC counting — that lives in tokei.
2. **`analysis/loc_report.sh`** — bash pipeline pinned to `tokei 14.0.0` (via `analysis/tokei_version.txt`, verified on startup). Runs tokei per source-set directory, extracts Kotlin `code` LOC and file counts, joins with the sharing report, and writes `analysis/data/sharing_report_with_loc.csv`.
3. **`python analysis/figures.py`** — pandas + matplotlib. Implements the **layer × platform-set LOC heatmap** (headline §4.2 figure) and stubs the other planned figures for Phase 4.

### The `platform_set` column — reach-aware aggregation

The naïve `source_set` name is **ambiguous**: `commonMain` of a full-platform KMP module reaches all 6 platforms including the JVM backend, while `commonMain` of a frontend-only KMP module reaches only the 5 frontends. Aggregating by raw source-set name would conflate these two different reach classes and overstate cross-tier (FE↔BE) sharing.

The Gradle task derives a stable `platform_set` label from three inputs, encoded in `PlatformReach.kt`:

- **Plugin family** — full-platform / frontend / backend / android-app. Snag convention plugins stack, so the *most specific* applied plugin wins.
- **Module platform classification** — the path-derived platform (`fe`, `be`, `common`, or blank). A module with `platform=fe` is architecturally scoped to frontend even if it applies the full-platform base plugin, because the category+platform direction rules forbid backend modules from depending on it. The path classification therefore narrows the effective family.
- **Source set name** — mapped to the concrete subset of platforms given the effective family.

The 16-label vocabulary is `all`, `frontend`, `nonAndroid_shared`, `nonWeb_shared`, `nonAndroid_fe`, `nonJvm`, `nonWeb_fe`, `mobile`, `web`, `jvm_shared`, `backend`, `jvm_desktop`, `android`, `ios`, `js`, `wasmJs`.

### Smoke-run results (on `main` + sync refactor from #224)

190/190 modules covered. 310 `(module × source set)` rows. 31,773 Kotlin LOC in production.

**Per-layer headline table (production only):**

| Layer              |   Total |    all | frontend | backend |
|--------------------|--------:|-------:|---------:|--------:|
| `business`         |     337 |    337 |        0 |       0 |
| `app` (hex)        |   6,931 |    136 |    4,068 |   2,100 |
| `ports`            |     550 |      0 |      391 |     159 |
| `driving`          |  13,341 |      0 |   10,218 |   1,574 |
| `driven`           |   5,594 |      0 |    3,209 |   2,181 |
| `contract`         |     267 |    267 |        0 |       0 |
| `core`             |     446 |    169 |      243 |       7 |
| `lib`              |   2,940 |     22 |    2,064 |     428 |
| `infra`            |     255 |     49 |      119 |      87 |
| `app (top-level)`  |   1,112 |      0 |      526 |     452 |
| **Total**          |  31,773 |    980 |   20,838 |   6,988 |

(The `feat (other)` residual row has dropped to zero and is no longer in the table. After PR #226 the cross-feature shared modules at `:featuresShared:database:*` and `:featuresShared:storage:*` have an explicit `driven` hex layer, so their 436 LOC now land cleanly in the `driven` row — the `driven` total moved from 5,158 to 5,594 and the `backend` column dropped by 40 LOC due to the introduction of an `encapsulation=api` split that reclassifies part of the shared DB code. The `feat (other)` label is kept in `figures.py`'s `LAYER_ORDER` as a regression tripwire.)

The **headline** for the thesis's §4.2 story: `all` (truly cross-platform, FE+BE) = 980 LOC ≈ 3.1 % of production. `frontend` (5-platform, FE-only) = 20,838 ≈ 65.6 %. The 19× gap between naive-aggregation numbers and reach-aware numbers is visible at a glance.

**Three observations the thesis highlights:**
- **`business` + `contract` are 100 % cross-tier shared.** 604 LOC of pure domain models and wire contracts land in the `all` (6-platform) bucket.
- **`app` (hex layer) is structurally symmetric but code-wise per-side.** 136 LOC of shared app-layer interfaces sit in `all`, but the 4,068 + 2,100 = 6,168 LOC of implementations are per-side. The hexagonal symmetry is an architectural invariant, not a code-reuse one — the NS Action Version Transparency story.
- **The `driving` layer has the biggest FE↔BE asymmetry:** 10,218 LOC in frontend vs 1,574 LOC in backend, zero in `all`. Expected cost of tying driving code to UI / HTTP frameworks (Compose Multiplatform on FE, ktor on BE) — and it is the single biggest opportunity for a thesis-flagged future refactor (driving-layer source-set split) identified by the Phase 0 Wear OS experiment.

**Bug flagged and fixed before findings were published.** The first revision of the heatmap aggregated by raw source-set name (`commonMain` etc.) and would have reported ~21,818 LOC as "shared", conflating the 980 LOC of true FE↔BE sharing with the 20,838 LOC of frontend-only sharing — overstating cross-tier reuse by roughly **19×**. The fix introduces the `platform_set` column above.

### Scope boundaries — intentionally out of Phase 1

- **No §4.2 Czech prose** — chapter text is drafted in Phase 4 once numbers stabilize.
- **No ripple classifier** — that's Phase 2.
- **No figures beyond the headline heatmap** — Phase 4 fills in the stubs.
- **No LOC counting inside the Gradle task** — tokei owns LOC.
- **No changes to any shared module** — the only production-code touch is one-line plugin application in the root `build.gradle.kts`.

### Unit Tests

- `build-logic/.../analysis/SharingReportRowBuilderTest.kt` — **60 `kotlin.test` cases** covering:
  - Every `ModuleIdentity` subtype: `CoreModule`, `LibModule`, `FeatModule`, `AppModule`, `InfraModule`
  - Every distinct field combination in `settings.gradle.kts`
  - Every `platform_set` label
  - **Regression: stacked-plugin trap** — `{multiplatform, frontend.multiplatform, driving.frontend.multiplatform}` must classify as FRONTEND, not FULL.
  - **Regression: fe/be path override** — a `:lib:X:fe:api` module applying the full-platform base plugin must still report `frontend` / `jvm_desktop` reach.
  - Edge cases: empty source-set list, non-canonical plugin filtered out, source-set dir round-trip, android-app marker classification.

### Manual Testing

- [x] `./gradlew :build-logic:test` — all 60 builder tests green
- [x] `./gradlew sharingReport` — 310 rows, 190 unique modules
- [x] `./gradlew check` — full project still builds and tests pass (post-rebase on sync refactor)
- [x] `analysis/loc_report.sh` — exits 0, tokei 14.0.0 version check passes
- [x] `python analysis/figures.py` — produces `fig_4_2_layer_platform_set_heatmap.pdf` + `.png`
- [x] Full pipeline **idempotent** — rerun produces byte-identical CSV

### Reproduction (fresh clone)

```bash
brew install tokei
./gradlew sharingReport
python3 -m venv analysis/.venv && source analysis/.venv/bin/activate
pip install -r analysis/requirements.txt
./analysis/loc_report.sh
python analysis/figures.py
```

## References

- **Depends on #226** — features-shared restructure. #226 itself depends on #224 (sync refactor). Merge chain: `main ← #224 ← #226 ← #223`.
- **Follow-up tracker: #225** — include SQLDelight `.sq` files (~400 LOC in `:featuresShared:database:fe:driven:api`) in the analysis. Out of scope for this PR.
- Evaluation plan: `analysis/thesis-evaluation-plan.md` (included — Phase 1 marked complete with findings inline)
- Phase 0 Wear OS feasibility log: `analysis/wearos_experiment.md`
- Existing infrastructure reused: `build-logic/.../architecture/ModulePathParser.kt`, `ModuleIdentity.kt`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: analysis/figures.py [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+                continue
+            color = "white" if val >= threshold else "black"
+            ax.text(j, i, f"{val:,}", ha="center", va="center", fontsize=8, color=color)
+
+    col_totals = data.sum(axis=0)
+    row_totals = data.sum(axis=1)
+    grand_total = int(data.sum())
+
+    col_labels = [
+        f"{c}\n({PLATFORM_SET_REACH_COUNT.get(c, 0)}p)\n{int(t):,}"
```
Comment (timotejAdamec) [inline analysis/figures.py]:
```suggestion
        f"{int(t):,} ({PLATFORM_SET_REACH_COUNT.get(c, 0)}p) {c}"
```
